### PR TITLE
doc work + type hints + refactor `simplify.simplify_{singletons, pairs}()`, etc.

### PR DESCRIPTION
**Pull request by jGaboardi**
This PR:
* does doc work + type hints + refactor + comments `simplify.simplify_pairs()`
* better explanation & comments – doc work refactor `simplify.simplify_singletons()`
* some DRYing & breaking out of helper funcs
* xref: 
  * https://github.com/uscuni/sgeop/issues/20
  * https://github.com/uscuni/sgeop/issues/21
* Questions regarding `non-planar clusters of value 2`
  * Q1: [`simplify@L351`](077b92ec5fb342ab222ccf9328c5a33c1bde0b6a28bd285e09ec42a51c04d73eR351)
  * Q2: [`simplify@L355`](https://github.com/uscuni/sgeop/pull/116/files#diff-077b92ec5fb342ab222ccf9328c5a33c1bde0b6a28bd285e09ec42a51c04d73eR353)

### CODE REVIEW

File: sgeop/simplify.py [hunk 1]
```diff
@@ -198,162 +248,175 @@ def simplify_singletons(
...
-        new_roads = pd.concat(
-            [cleaned_roads, new],
-            ignore_index=True,
-        )
-        new_roads = remove_false_nodes(
-            new_roads[~(new_roads.is_empty | new_roads.geometry.isna())],
-            aggfunc={"_status": _status},
-        )
+        new_roads = pd.concat([cleaned_roads, new], ignore_index=True)
+        _empties = new_roads[~(new_roads.is_empty | new_roads.geometry.isna())]
```
Comment (martinfleis) [inline sgeop/simplify.py:264]: These are not empty but non-empty

File: sgeop/simplify.py [hunk 2]
```diff
@@ -198,162 +248,175 @@ def simplify_singletons(
...
-    artifacts.loc[artifacts.index[np.unique(a_idx)], "non_planar"] = True
+    # Filter artifacts caused by non-planar intersections
+    artifacts = _identify_non_planar(artifacts, roads)
 
-    artifacts["non_planar_cluster"] = artifacts.apply(
-        lambda x: sum(artifacts.loc[artifacts["comp"] == x.comp]["non_planar"]), axis=1
-    )
+    # Identify non-planar clusters
+    _id_np = lambda x: sum(artifacts.loc[artifacts["comp"] == x.comp]["non_planar"])  # noqa: E731
+    artifacts["non_planar_cluster"] = artifacts.apply(_id_np, axis=1)
```
Comment (martinfleis) [inline sgeop/simplify.py:340]: Sometimes I wonder if refactors of this sort are worth it. I don't consider the original less readable (maybe more), and it does not require linting exception.

File: sgeop/simplify.py [hunk 3]
```diff
@@ -198,162 +248,175 @@ def simplify_singletons(
...
-    )
+    # Isolate planar artifacts
+    _planar_grouped = artifacts_planar.groupby("comp")[artifacts_planar.columns]
+    _solutions = _planar_grouped.apply(get_solution, roads=roads)
+    artifacts_w_info = artifacts.merge(_solutions, left_on="comp", right_index=True)
+
+    # Isolate non-planar clusters of value 2
+    # ---> MARTIN ---> what exactly does this mean? ###################################
+    _np_clust_2 = np_clusters["non_planar_cluster"] == 2
+    # ---> MARTIN ---> what does does it mean to be 'under non-planar'? ###############
```
Comment (martinfleis) [inline sgeop/simplify.py:353]: You have an artifact under the highway, for example. You need to ignore the highway to solve the artifact below.

File: sgeop/simplify.py [hunk 4]
```diff
@@ -198,162 +248,175 @@ def simplify_singletons(
...
+        # Isolate artifacts to process as clusters
         for_skeleton = artifacts_w_info.query("solution == 'skeleton'")
+
+    # Otherwise instantiate artifact containers as empty
     else:
         merged_pairs = pd.DataFrame()
-        first = pd.DataFrame()
-        second = pd.DataFrame()
+        _1st = pd.DataFrame()
+        _2nd = pd.DataFrame()
```
Comment (martinfleis) [inline sgeop/simplify.py:397]: why??

**Review by martinfleis — COMMENTED** [review]
I think that you may be overdoing it occasionally...

**Comment by jGaboardi** [comment]
> I think that you may be overdoing it occasionally...

Agreed. Was thinking that when I was pushing this up.

File: sgeop/simplify.py (hunk 2 above)
Comment (jGaboardi) [reply sgeop/simplify.py:340]: While I think this is more readable, I completely understand your view. I have no problem reverting.

File: sgeop/simplify.py (hunk 1 above)
Comment (jGaboardi) [reply sgeop/simplify.py:264]: Good point

File: sgeop/simplify.py (hunk 4 above)
Comment (jGaboardi) [reply sgeop/simplify.py:397]: More readable for me, but I have no problem reverting.

**Comment by jGaboardi** [comment]
So how do we feel if I revert most of the non-function refactors? I probably went too far down the rabbit hole of this.

**Comment by martinfleis** [comment]
Leave it as it is but avoid doing that next time :)

**Comment by jGaboardi** [comment]
Fair enough.

**Comment by jGaboardi** [comment]
@martinfleis Is this OK to merge?

**Pull request merged**